DasharoSystemFeaturesUiLib: BtG Status Menu

mkopec opened the pull request.

mkopec committed:
DasharoSystemFeaturesUiLib: Add Boot Guard info to ME menu

Signed-off-by: Micha¿ Kope¿ <michal.kopec@3mdeb.com>

mkopec commented:
![image](https://github.com/user-attachments/assets/42296eb4-0c48-4d13-a428-84e7bf5d7d63)
Looks like this

miczyg1 commented:
@mkopec looks generalyl good, but you went with all integers. I hoped you would choose some `oneof`s, e.g.:
`TPM Type                              <PTT>`

In some places it makes more sense to do so. Also the last one is the TPM startup locality, 0 means startup from locality 3 (so 0/1 doesn't reflect the locality itself but kind of inverted logic), os it would be great to have 

`TPM Startup done from                    <Locality 3>`
`TPM Startup done from                    <Locality 0>`

We don't need integers for parsed values if we have the raw HEX of the register. My suggestion:

```
NEM Enabled                             <Yes/No>
TPM Type                                   <No TPM/dTPM 1.2/dTPM2.0/PTT>
TPM Success                              <Yes/No>
...
Reset Secrets Protection          <No/Yes>
```

Pretty much every boolean could be made Yes/No or No/Yes (if inverted logic applies).

Fields which would love to have a string descriptions too (like TPM Type):

- S-CRTM status
- Memory Scrubbing policy
- TXT profile
- Backup Action

mkopec commented:
I've changed from numerics to oneofs and added handling for inverted logic

![btg_status](https://github.com/user-attachments/assets/510d287a-f6d5-4bd2-bb7c-507d6ac58b49)

miczyg1 commented:
There should be `TXT disabled by policy` (inverted). If one would search for `TXT enabled by policy` phrase, it would be more difficult to find.

The rest looks good. Will take a look at the code later.

miczyg1 commented on DasharoModulePkg/Library/DasharoSystemFeaturesUiLib/DasharoSystemFeaturesVfr.vfr:
```diff
@@ -233,6 +233,530 @@ formset
              option text = STRING_TOKEN(STR_ME_MODE_DISABLE_HAP), value = ME_MODE_DISABLE_HAP, flags = 0;
           endif;
       endoneof;
+      grayoutif TRUE;
```
Add an `disableif` for PCD that will control the visibility of this status information, above the `grayoutif`.

mkopec replied:
https://github.com/Dasharo/edk2/pull/248/commits/c82ebb6d73b8275b12baf834bfca5c1e237c3903

miczyg1 commented on DasharoModulePkg/Library/DasharoSystemFeaturesUiLib/DasharoSystemFeatures.c:
```diff
@@ -646,6 +647,18 @@ DasharoSystemFeaturesCallback (
       Value->u8 = FixedPcdGet8(PcdCpuMaxTemperature) - Private->DasharoFeaturesData.CpuThrottlingOffset;
       break;
+    case IBG_SACM_INFO_MSR_QUESTION_ID:
+      Value->u64 = AsmReadMsr64(0x13A);
```
Reading an inexistent MSR may cause trouble (i.e. exception) on some platforms. I think GLK and SNB/IVB does not have this MSR. Thus the `disableif` suggested in other comment, so that this will never get called on platforms that are not supposed to support it (but still may have ME disable/enable option).

mkopec replied:
https://github.com/Dasharo/edk2/pull/248/commits/c82ebb6d73b8275b12baf834bfca5c1e237c3903

miczyg1 commented on DasharoModulePkg/Library/DasharoSystemFeaturesUiLib/DasharoSystemFeaturesVfr.vfr:
```diff
@@ -233,6 +233,530 @@ formset
+                 option text = STRING_TOKEN(STR_YES), value = 1, flags = 0;
+        endoneof;
+
+        oneof varid   = FeaturesData.IbgStatus.BootStatus.Bits.TxtDisPol,
+                prompt  = STRING_TOKEN(STR_IBG_STATUS_BOOT_STATUS_TXT_ENABLED_BY_POLICY),
+                help    = STRING_TOKEN(STR_IBG_STATUS_BITFIELD_HELP),
+                flags   = DISPLAY_UINT_HEX | INTERACTIVE,
+                minimum = 0,
+                maximum = 1,
+
+                 option text = STRING_TOKEN(STR_YES), value = 0, flags = 0;
+                 option text = STRING_TOKEN(STR_NO), value = 1, flags = 0;
+        endoneof;
```
```suggestion
        oneof varid   = FeaturesData.IbgStatus.BootStatus.Bits.TxtDisPol,
                prompt  = STRING_TOKEN(STR_IBG_STATUS_BOOT_STATUS_TXT_DISABLED_BY_POLICY),
                help    = STRING_TOKEN(STR_IBG_STATUS_BITFIELD_HELP),
                flags   = DISPLAY_UINT_HEX | INTERACTIVE,
                minimum = 0,
                maximum = 1,

                 option text = STRING_TOKEN(STR_NO), value = 0, flags = 0;
                 option text = STRING_TOKEN(STR_YES), value = 1, flags = 0;
        endoneof;

```

miczyg1 commented on DasharoModulePkg/Library/DasharoSystemFeaturesUiLib/DasharoSystemFeaturesStrings.uni:
```diff
@@ -233,3 +233,95 @@
+#string STR_IBG_STATUS_BOOT_STATUS_MEM_POWER_DOWN         #language en-US "    Memory Power Down executed"
+#string STR_IBG_STATUS_BOOT_STATUS_BTG_FAILED             #language en-US "    Boot Guard Failed"
+#string STR_IBG_STATUS_BOOT_STATUS_BIOS_TRUSTED           #language en-US "    BIOS is Trusted"
+#string STR_IBG_STATUS_BOOT_STATUS_TXT_ENABLED_BY_POLICY #language en-US "    TXT is enabled by policy"
```
```suggestion
#string STR_IBG_STATUS_BOOT_STATUS_TXT_DISABLED_BY_POLICY #language en-US "    TXT disabled by policy"

```

mkopec committed:
BtG status menu: Keep inverted logic for TxtDisPol

Co-authored-by: Michał Żygowski <michal.zygowski@3mdeb.com>

mkopec committed:
DasharoModulePkg: add PCD to control Intel CBnT status visibility

Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>
